src/main/kotlin/com/openlattice/linking/matching/SocratesMatcher.kt

```diff
@@ -82,56 +97,83 @@ class SocratesMatcher(model: MultiLayerNetwork, private val fqnToIdMap: Map<Full
             block: Pair<EntityDataKey, Map<EntityDataKey, Map<UUID, Set<Any>>>>
     ): Pair<EntityDataKey, MutableMap<EntityDataKey, MutableMap<EntityDataKey, Double>>> {
         val sw = Stopwatch.createStarted()
-        val model = localModel.get()
 
+        val model = localModel.get()
         val entityDataKey = block.first
         val entities = block.second
 
-        val extractedEntities = entities.mapValues { extractProperties(it.value) }
 
-        val matchedEntities = extractedEntities.mapValues {
-            val entity = it.value
-            extractedEntities
-                    .mapValues { computeScore(model, entity, it.value) }
-                    .toMutableMap()
+        // extract properties and features for all entities in block
+        val extractedFeatures = entities.mapValues {
+            val selfProperties = extractProperties(it.value)
+            entities.mapValues {
+                val otherProperties = extractProperties(it.value)
+                extractFeatures(selfProperties, otherProperties)
+            }
+        }
+
+        // transform features to matrix and compute scores
+        val featureMatrix = extractedFeatures.flatMap { (entityDataKey1, features) ->
+            listOf(features.map {
```

Comment:
I'm a little confused by the creation of the inner list, followed by the subsequent flatten. That seems redundant.

Reply:
i.e ```extractedFeature.flatMap{ ( edk, f ) ->  feature.map { it.value[0] } }``` would do the same thing. Am I missing something?

Reply:
Nope, that is redundant indeed :)